src/GNDStk/gnds-convert-to-json.hpp

```diff
@@ -0,0 +1,134 @@
+
+/*
+Summary of the functions in this file:
+
+namespace detail {
+   1. prefix(unsigned long)
+
+   2. Node2json(gnds::Node, nlohmann::json, unsigned long)
+      ...uses (1)
+}
+
+3. convert(gnds::Tree, gnds::json)
+   ...uses (2)
+
+4. convert(gnds::xml,  gnds::json)
+   ...uses (3)
+
+5. convert(gnds::json, gnds::json)
+*/
+
+
+
+// -----------------------------------------------------------------------------
+// Helpers
+// -----------------------------------------------------------------------------
+
+namespace detail {
+
+// prefix
+inline std::string prefix(const unsigned long n)
+{
+   std::ostringstream oss;
+   oss << std::setfill('0') << std::setw(12) << n;
+   return oss.str();
+}
+
+
+
+// Node2json
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool Node2json(
+   const gnds::Node<MCON,CCON> &node,
+   nlohmann::json &j,
+   unsigned long &kwdcount
```

Comment:
No worries. That was actually something that I put in for debugging purposes, and the plan is to remove it, except perhaps optionally in a "debug mode" or the like, when I complete the JSON work.

Re: debugging. Although JSON (and GNDS in general) doesn't have a general ordering concept, having an optional "put an ordering prefix on everything" feature proved to be useful for vetting the initial JSON capability. In short: I wrote some stress tests that read each of the several thousand .xml format GNDS files, outputted in JSON, read back in as JSON, then wrote back out as XML, then finally checked that we ended up with exactly the same thing. Checking for the "same thing" is tougher if everything gets reordered - even if ordering, per se, isn't part of a format's specification. Such a check can certainly be done in other ways, but I decided that this way was easy to write, and would do the trick.

Generally speaking, then, you'll sometimes see this sort of thing in the code I write. It's there for vetting/debugging, not because of any misunderstanding about the nature of what I'm working with. And, sometimes, I'll leave the capability in the code (turned off by default, via a flag or conditional compilation), for easy use in the future if something goes wrong.

It's a bit like how modern auto manufacturers make plug-in ports (maybe just USB ports these days?) available for mechanics, or expert users a.k.a. drivers, to plug in diagnostic equipment if and when something goes wrong, and "talk" with the car's computer. The presence of the ports shouldn't be construed as meaning the car is built poorly, and the ports aren't meant to clutter up the car's internals to an extent that a mechanic would be confused by it being there. That's usually the basic philosophy that's going on when you see this sort of thing in my work.

Reply:
So, I see your reasoning for putting in `kwdcount`. We prefer to do this by writing tests—and keeping them. You state that you have written some stress tests. Why not write these tests in a way that can be captured and stored? You've spent some effort on writing them, let's keep that around.

I know you are not fond of the Test-Driven-Development, but this is a big reason why we do that. We don't want to write code then throw it away when the test is done.

Reply:
Given that there are lots of things to test, and that I have indeed written more elsewhere... One question I'd have is what we'd like to do, given the way things are supposed to be done on this project, to keep the individual test files under control, size-wise. The .cpp files in Tree/test/ comprise nearly 2000 lines right now, and additional tests for JSON will need to be in there soon. These would no doubt end up in the read and write tests. Is there any particular convention you'd prefer instead; say, maybe a read.json.test.cpp or the like? Tests are split up according to the particular function name (read, write, ctor for constructor, etc.), but when variations exist withing that context (read XML, read JSON, etc.), how do you prefer that it be handled?

Reply:
I have indeed already put the tests in question in different files per the style guide. Tests of read() are in read.test.cpp, for instance, and tests of write() are in write.test.cpp. I was thinking in terms of, "what if read() and write() have various ways of calling them [read XML, read JSON, read HDF, etc.] and we don't want read.test.cpp and write.test.cpp to get longer." So, is there a style you use for getting a finer scale: read.xml.test.cpp, read.json.test.cpp, etc.? I can certainly put more tests into the files that are named for the functions they test, but thought an even finer scale might be reasonable.

Reply:
This kind of behavior isn't specified in the Style Guide. I think what you have described is a good approach. Why are these tests so long? Is it because there is lots of XML/JSON hard-coded?

Reply:
It's because my works tend to be feature dense, with lots of variations in how to call and use things. So, lots of things to test. You never know when what seems like a tiny testing variation could trigger a latent issue of some kind. I can't remember the last time a defect was reported in any of my work, but it's software, so anything could surface at any time.

Reply:
kwdcount isn't there any longer. The basic thing I'd written it to help with, can now be accomplished by using the node sort() function.